File: packages/local_auth/ios/Classes/Biology.m
//
//  Biology.m
//  local_auth
//
//  Created by 李广斌 on 2020/9/24.
//

#import "Biology.h"
#import <LocalAuthentication/LocalAuthentication.h>

@implementation Biology

+(void) biologyUnLockTips:(NSString *)tips negativeBtn:(NSString *)negativeBtn positiveBtn:(NSString *)positiveBtn unLockSuccess:(void (^)(void)) unLockSuccess unLockFail:(void (^)(NSInteger)) unLockFail unLockError:(void (^)(NSInteger)) unLockError {
    
    LAContext *authenticationContext = [[LAContext alloc] init];
    
    authenticationContext.localizedFallbackTitle = positiveBtn;
    if (@available(iOS 10.0, *)) {
        authenticationContext.localizedCancelTitle = negativeBtn;
    } else {
    }
    
    NSError *error = nil;
    BOOL isSupport = [authenticationContext canEvaluatePolicy: LAPolicyDeviceOwnerAuthentication error:&error];
    
    if (isSupport) {
        
        [authenticationContext evaluatePolicy: LAPolicyDeviceOwnerAuthentication localizedReason:tips reply:^(BOOL success, NSError * _Nullable error) {

            [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                if (success) {
                    unLockSuccess();
                }
                else{
                    unLockFail(error.code);
                }
            }];
        }];
    }
    else {
        
        [[NSOperationQueue mainQueue] addOperationWithBlock:^{
            unLockError(error.code);
        }];
    }
}

+(NSInteger)isSupport {
    LAContext *context = [[LAContext alloc] init];
    NSError *error = nil;
    [context canEvaluatePolicy: LAPolicyDeviceOwnerAuthentication error:&error];
    
    if (error == nil) {
        return 0;
    }
    else {
        return error.code;
    }
}

+(NSInteger)getWhitchBiology {
    
    LAContext *context = [[LAContext alloc] init];
    NSError *error = nil;
    [context canEvaluatePolicy: LAPolicyDeviceOwnerAuthentication error:&error];
    if (@available(iOS 11.0, *)) {
        return [context biometryType];
    } else {
        return 1;
    }
}

///  去设置权限
+(void) setBiologyOkTitle1:(NSString *)title1 negativeBtn:(NSString *)negativeBtn positiveBtn:(NSString *)positiveBtn onNegative:(void (^)(void))onNegative onPositive:(void (^)(void))onPositive{
    
    UIAlertController *one = [UIAlertController alertControllerWithTitle:title1 message:@"" preferredStyle: UIAlertControllerStyleAlert];
    
    UIAlertAction *negative = [UIAlertAction actionWithTitle:negativeBtn style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        onNegative();
    }];
    
    UIAlertAction *positive = [UIAlertAction actionWithTitle:positiveBtn style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        [self systemSeting];
    }];
    
    [one addAction: negative];
    [one addAction:positive];
    
    [[[UIApplication sharedApplication] keyWindow].rootViewController presentViewController:one animated:YES completion:nil];
}


///  去设置权限
/// 跳转到系统设置主页
+(void)systemSeting {
    NSURL *url = [[NSURL alloc] initWithString: UIApplicationOpenSettingsURLString];
    if (@available(iOS 10.0, *)) {
        [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:^(BOOL success) {
        }];
    } else {
        [[UIApplication sharedApplication] openURL:url];
    }
}


+(void) setBiologyshouldOpenTitle2:(NSString *)title2 negativeBtn:(NSString *)negativeBtn onNegative:(void (^)(void))onNegative{
    
    UIAlertController *one = [UIAlertController alertControllerWithTitle:title2 message:@"" preferredStyle: UIAlertControllerStyleAlert];
    
    UIAlertAction *negative = [UIAlertAction actionWithTitle:negativeBtn style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        onNegative();
    }];
    
    [one addAction: negative];
    
    [[[UIApplication sharedApplication] keyWindow].rootViewController presentViewController:one animated:YES completion:nil];
}


+(void) biologyOutTitle3:(NSString *)title3 negativeBtn:(NSString *)negativeBtn positiveBtn:(NSString *)positiveBtn onNegative:(void (^)(void))onNegative onPositive:(void (^)(void))onPositive{
    
    UIAlertController *one = [UIAlertController alertControllerWithTitle:title3 message:@"" preferredStyle: UIAlertControllerStyleAlert];
    
    UIAlertAction *negative = [UIAlertAction actionWithTitle:negativeBtn style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        onNegative();
    }];
    
    UIAlertAction *positive = [UIAlertAction actionWithTitle:positiveBtn style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        onPositive();
    }];
    
    [one addAction: negative];
    [one addAction:positive];
    
    [[[UIApplication sharedApplication] keyWindow].rootViewController presentViewController:one animated:YES completion:nil];
}


@end

File: packages/local_auth/ios/Classes/Biology.swift
//
//  Biology.swift
//
//  Created by 李广斌 on 2020/9/22.
//  Copyright © 2019年 Local_auth. All rights reserved.
//

import UIKit
import LocalAuthentication

@objc public class Biology: NSObject {
    
    @objc public static func biologyUnLock(tip:String = " ",sensitiveTransaction:Int,negativeBtn:String,positiveBtn:String,unLockSuccess:@escaping (()->()),unLockFail:@escaping ((_ failCode:Int)->()),unLockError:@escaping ((_ code:Int)->())){
        let authenticationContext = LAContext()
        var err:NSError?

        authenticationContext.localizedFallbackTitle = positiveBtn

        if #available(iOS 10.0, *) {
            authenticationContext.localizedCancelTitle = negativeBtn
        } else {
            
        }
        let isTouchIdAvailable = authenticationContext.canEvaluatePolicy(.deviceOwnerAuthentication,error: &err)
        
        if isTouchIdAvailable{
            authenticationContext.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: tip) { (success, error) in
                OperationQueue.main.addOperation {
                    if success{
                        unLockSuccess()
                    }else{
                        let erroring = (error as NSError?)
                        unLockFail(erroring?.code ?? 0)
                    }
                }
            }
        }else{
            OperationQueue.main.addOperation {
                unLockError(err?.code ?? 0)
            }
        }
    }
    
    @objc public static func isSupport()->Int{
        let authenticationContext = LAContext()
        var err:NSError?
        _ = authenticationContext.canEvaluatePolicy(.deviceOwnerAuthentication,error: &err)
        
        return err?.code ?? 0
    }
    
    @objc public static func getWhitchBiology()->Int{
        let authenticationContext = LAContext()
        var err:NSError?
        _ = authenticationContext.canEvaluatePolicy(.deviceOwnerAuthentication,error: &err)
        
        if #available(iOS 11.0, *) {
            return authenticationContext.biometryType.rawValue
        }
        return 1
    }
    
    @objc public static func setBiologyOk(title1:String,negativeBtn:String,positiveBtn:String,onPositive:(()->())?,onNegative:(()->())?){
        let one = UIAlertController(title: title1, message: "", preferredStyle: UIAlertController.Style.alert)
        
        let positiveBtn = UIAlertAction(title: "去开启", style: UIAlertAction.Style.default, handler: { (action) in
            onPositive?()
            jumpToSystemSeting()
            
        })
        let negativeBtn = UIAlertAction(title: "取消", style: UIAlertAction.Style.default, handler: { (action) in
            onNegative?()
        })
        
        one.addAction(negativeBtn)
        one.addAction(positiveBtn)
        
        UIApplication.shared.keyWindow?.rootViewController?.present(one, animated: true, completion: nil)
    }
    /// 跳转到系统设置主页
    fileprivate static func jumpToSystemSeting() {
        let settingUrl = URL(string: UIApplication.openSettingsURLString)
        if let url = settingUrl, UIApplication.shared.canOpenURL(url) {
            UIApplication.shared.openURL(url)
        }
    }
    
    @objc public static func setBiologyshouldOpen(title2:String,negativeBtn:String,positiveBtn:String,onNegative:(()->())?){
        let one = UIAlertController(title: title2, message: "", preferredStyle: UIAlertController.Style.alert)
        
        let negativeBtn = UIAlertAction(title: negativeBtn, style: UIAlertAction.Style.default, handler: { (action) in
            onNegative?()
        })
        
        one.addAction(negativeBtn)
        UIApplication.shared.keyWindow?.rootViewController?.present(one, animated: true, completion: nil)
    }
    
    @objc public static func biologyOut(title3:String,negativeBtn:String,positiveBtn:String,onPositive:(()->())?,onNegative:(()->())?){
        
        let one = UIAlertController(title: title3, message: "", preferredStyle: UIAlertController.Style.alert)
        
        let negativeBtn = UIAlertAction(title: negativeBtn, style: UIAlertAction.Style.default, handler: { (action) in
            onNegative?()
        })
        
        let positiveBtn = UIAlertAction(title: positiveBtn, style: UIAlertAction.Style.default, handler: {(action) in
            onPositive?()
            /// 密码支付回调
        })
        one.addAction(negativeBtn)
        one.addAction(positiveBtn)
        UIApplication.shared.keyWindow?.rootViewController?.present(one, animated: true, completion: nil)
    }
    
}
